### PR TITLE
Combine site specific scripts from component updater

### PR DESCRIPTION
**Pull request by sangwoo108**
Also fixes bugs in Android.

Resolves https://github.com/brave/brave-browser/issues/28130

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/playlist/browser/media_detector_component_manager.cc [hunk 1]
```diff
@@ -108,21 +160,34 @@ bool MediaDetectorComponentManager::ShouldHideMediaSrcAPI(
...
     constexpr std::string_view kPlaceholder =
         "const siteSpecificDetector = null";
     auto pos = detector_script.find(kPlaceholder);
+    if (pos == std::string::npos) {
+      // Reportedly, in some environment(e.g. Android release), the js script
+      // resource could be minified by removing white spaces.
+      constexpr std::string_view kPlaceholderWithoutWhitespace =
+          "const siteSpecificDetector=null";
+      pos = detector_script.find(kPlaceholderWithoutWhitespace);
+    }
```
Comment (sangwoo108) [inline components/playlist/browser/media_detector_component_manager.cc:183]: @deeppandya , this is what we found during the call

File: components/playlist/browser/playlist_service.cc [hunk 2]
```diff
@@ -516,7 +516,7 @@ void PlaylistService::OnThumbnailDownloaded(const std::string& id,
   const auto* value = prefs_->GetDict(kPlaylistItemsPref).FindDict(id);
   DCHECK(value);
   auto playlist_item = ConvertValueToPlaylistItem(*value);
-  playlist_item->thumbnail_path = GURL(path.AsUTF8Unsafe());
+  playlist_item->thumbnail_path = GURL("file://" + path.AsUTF8Unsafe());
```
Comment (sangwoo108) [inline components/playlist/browser/playlist_service.cc:519]: @deeppandya and this one too

**Review by sangwoo108 — COMMENTED** [review]
@deeppandya , I've addressed what we ran into during the call. Please let me know it's not proper solution.

File: components/playlist/browser/media_detector_component_manager.cc [hunk 3]
```diff
@@ -19,8 +20,38 @@ namespace playlist {
...
+}
+
+const base::flat_map<ScriptName, std::string>& GetLocalScriptMap() {
+  static const base::NoDestructor local_resource_map(([]() {
+    const auto& rb = ui::ResourceBundle::GetSharedInstance();
+    base::flat_map<ScriptName, std::string> local_resource_map = {
+        {GetBaseScriptName(),
+         rb.LoadDataResourceString(IDR_PLAYLIST_MEDIA_DETECTOR_JS)},
+        {FILE_PATH_LITERAL("youtube.com.js"),
+         rb.LoadDataResourceString(IDR_PLAYLIST_MEDIA_DETECTOR_YOUTUBE_JS)},
```
Comment (goodov) [inline components/playlist/browser/media_detector_component_manager.cc:49]: cc @atuchin-m

Comment (atuchin-m) [reply components/playlist/browser/media_detector_component_manager.cc:49]: Please remove caching here.
See https://bravesoftware.slack.com/archives/C7VLGSR55/p1675100805299529

File: components/playlist/browser/media_detector_component_manager.cc [hunk 4]
```diff
@@ -29,11 +60,16 @@ std::string ReadScript(const base::FilePath& path) {
...
 
-const std::string& GetLocalScript() {
-  static const std::string kScript =
-      ui::ResourceBundle::GetSharedInstance().LoadDataResourceString(
-          IDR_PLAYLIST_MEDIA_DETECTOR_JS);
-  return kScript;
+base::flat_map<ScriptName, std::string> ReadScriptsFromComponent(
+    base::flat_set<base::FilePath> files) {
+  base::flat_map<ScriptName, std::string> script_map;
+  for (const auto& path : files) {
```
Comment (atuchin-m) [inline components/playlist/browser/media_detector_component_manager.cc:66]: @sangwoo108 
What it the typical number/size of these files?

File: components/playlist/browser/media_detector_component_manager.cc [hunk 5]
```diff
@@ -69,20 +105,39 @@ void MediaDetectorComponentManager::RegisterIfNeeded() {
...
+void MediaDetectorComponentManager::OnGetScripts(
+    const MediaDetectorComponentManager::ScriptMap& script_map) {
+  if (script_map.empty()) {
+    LOG(ERROR) << __FUNCTION__ << " scripts are empty!";
     return;
   }
 
-  script_ = script;
+  DCHECK(script_map.count(GetBaseScriptName()));
+  script_ = script_map.at(GetBaseScriptName());
```
Comment (simonhong) [inline components/playlist/browser/media_detector_component_manager.cc:129]: nit: As we'll have base and site specific scripts, how about `base_script_`?

**Review by simonhong — APPROVED** [review]
++

**Comment by simonhong** [comment]
BTW, instead of handling multiple script files at browser side, I think CI can make that mapping and deliver it to browser?

**Comment by sangwoo108** [comment]
> I think CI can make that mapping and deliver it to browser?

Yeah, I'm thinking of making json file containing those information in playlist-component.

File: components/playlist/browser/media_detector_component_manager.cc (hunk 3 above)
Comment (sangwoo108) [reply components/playlist/browser/media_detector_component_manager.cc]: Great! Replaced with RawResource version and removed caching

File: components/playlist/browser/media_detector_component_manager.cc (hunk 4 above)
Comment (sangwoo108) [reply components/playlist/browser/media_detector_component_manager.cc:62]: For now we have 2 scripts, and they are 5kb and 2kb respectively.

File: components/playlist/browser/media_detector_component_manager.cc (hunk 5 above)
Comment (sangwoo108) [reply components/playlist/browser/media_detector_component_manager.cc]: That's a way better name. Changed.

File: components/playlist/browser/media_detector_component_manager.cc (hunk 3 above)
Comment (sangwoo108) [reply components/playlist/browser/media_detector_component_manager.cc]: Oops, calling RawResource doesn't work. I thought the script were not zipped but I seem to be wrong. Reverted to LoadDataResourceString.

File: components/playlist/browser/media_detector_component_manager.cc [hunk 6]
```diff
@@ -19,8 +20,34 @@ namespace playlist {
...
+
+base::flat_map<ScriptName, std::string> GetLocalScriptMap() {
+  const auto& rb = ui::ResourceBundle::GetSharedInstance();
+  return {
+      {GetBaseScriptName(),
+       std::string(rb.LoadDataResourceString(IDR_PLAYLIST_MEDIA_DETECTOR_JS))},
+      {FILE_PATH_LITERAL("youtube.com.js"),
+       std::string(
+           rb.LoadDataResourceString(IDR_PLAYLIST_MEDIA_DETECTOR_YOUTUBE_JS))},
+  };
```
Comment (sangwoo108) [inline components/playlist/browser/media_detector_component_manager.cc:50]: Removed cache

File: components/playlist/browser/media_detector_component_manager.cc (hunk 4 above)
Comment (sangwoo108) [reply components/playlist/browser/media_detector_component_manager.cc:62]: @atuchin-m said that if a file becomes bigger than 1mb, then we should revisit this.

Comment (atuchin-m) [reply components/playlist/browser/media_detector_component_manager.cc:62]: It's okay for now. If the total size becomes more that 1MB, please ping me.

File: components/playlist/browser/media_detector_component_manager.cc (hunk 3 above)
Comment (atuchin-m) [reply components/playlist/browser/media_detector_component_manager.cc]: Ok. Removing the cache in enough here.

**Pull request merged**